Adds Image and Text 25/75 molecule

anselmbradford opened the pull request:
Adds Image and Text 25/75 molecule.

## Additions

- Added templates and CSS for the Image and Text 25/75 molecule.

## Testing

- Added Image and Text 25/75 to `/landing-page-1/` and `/landing-page-2/`

## Review

- @KimberlyMunoz 
- @sebworks 
- @jimmynotjim 

## Screenshots

![screen shot 2015-10-06 at 1 23 20 pm](https://cloud.githubusercontent.com/assets/704760/10316485/732cc3f8-6c2d-11e5-897b-24a146827c18.png)

## Todos

- When I used `unit` to convert the bottom margin to ems it doubled in width from what I expected, but only on the h1 and h2. See https://github.com/cfpb/cfgov-refresh/compare/dfjr_382_image_and_text_25_75?expand=1#diff-871f1730644230967b426338ff60df0fR70 and https://github.com/cfpb/cfgov-refresh/compare/dfjr_382_image_and_text_25_75?expand=1#diff-871f1730644230967b426338ff60df0fR85
Any ideas? cc @Scotchester

sebworks commented on cfgov/v1/preprocessed/css/molecules/image-and-text-25-75.less:
```diff
@@ -0,0 +1,93 @@
+    &_image-container {
+
+        > img {
+            width: 130px;
```
should the widths be variables?

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/5ac8bf0799c8b21db7fdb25ce75a515997daa7da

Scotchester commented:
@anselmbradford That's happening because you're calculating the em value based on `@base-font-size-px`, (16px), not the actual font size of the element to which you're applying the margin.

To work around this, you'd need to have individual declarations for each heading size, and use the known value of their font size in the calculation, instead of `@base-font-size-px`.

anselmbradford commented:
> @anselmbradford That's happening because you're calculating the em value based on @base-font-size-px, (16px), not the actual font size of the element to which you're applying the margin.

Thanks @Scotchester that makes perfect sense now! What's our guideline again for when to use units and when to not? Would the unit conversion be required in this case at all?

Scotchester commented:
@anselmbradford Such guidelines are... murky? Nonexistent? I think that in the vast majority of cases, we want to use the em conversion. Reasons not to would include:

1. It's difficult/impossible to perform the calculation, because we don't know what the in-context font size is.
2. If we envision how things would look with a user-specified larger default font size (not a page zoomed in with full-page-zoom), would we still want the value to be hardcoded to a certain pixel value, or would we want it to scale with the text size? For example, I could see how we might always want our gutters between columns to stick at 30px, regardless of font size. But I think top/bottom margins and paddings should almost always scale, since the vertical whitespace rhythm affects readability.

Ultimately, because of #2, it's a judgment call you might sometimes make, but I think the default assumption should be making the em conversion.

Let me know if that didn't make sense.

jimmynotjim commented on cfgov/v1/preprocessed/css/molecules/image-and-text-25-75.less:
```diff
@@ -0,0 +1,93 @@
+h1.m-image-and-text-25-75_heading {
+    /* TODO: Margin is wider than expecting when applied to h1.
+    margin-bottom: unit( ( @grid_gutter-width * 2 ) / @base-font-size-px, em ); */
+    margin-bottom: 60px;
```
That's because you're using `@base-font-size` but the `h1` is 34px at > 600 screens and 26px < 600 screens. Same for above. There's actually a hack around this that I had to put a fix in for in cf-core because I broke it. By calling `.h1();` we expose a `@font-size` var that you could then use in place of `@base-font-size` and it adjusts itself automatically at the 600px bp.

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/517e2ad733b18b516a737618a6ce54f87f4a2958

jimmynotjim commented on cfgov/v1/preprocessed/css/molecules/image-and-text-25-75.less:
```diff
@@ -0,0 +1,93 @@
+    &_link {
+        /* TODO: Margin is wider than expecting when applied to h2.
+        margin-bottom: unit( @grid_gutter-width / @base-font-size-px, em ); */
+        margin-bottom: 30px;
```
For this one, you'd have to break out the heading, because it's a different font-size than the body and link

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/517e2ad733b18b516a737618a6ce54f87f4a2958

jimmynotjim commented on cfgov/v1/preprocessed/css/molecules/image-and-text-25-75.less:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    &_content::after {
+        content: '';
```
Don't we have a clearfix mixin?

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/a983212d0b72d169996734662bace56e8a58eeaa

KimberlyMunoz commented on cfgov/v1/jinja2/v1/_includes/molecules/image-and-text-25-75.html:
```diff
@@ -0,0 +1,29 @@
+{% import template as content %}
+<div class="m-image-and-text-25-75">
+    <div class="m-image-and-text-25-75_image-container">
+        <img src="{{ img_url }}">
```
We'll want either a decorative alt tag or a way to pass in alt tag text.

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/5ac8bf0799c8b21db7fdb25ce75a515997daa7da

jimmynotjim commented on cfgov/v1/preprocessed/css/molecules/image-and-text-25-75.less:
```diff
@@ -0,0 +1,93 @@
+
+    &_image-container {
+
+        > img {
```
In the GHE issue we discussed that there may be situations where this will be a bg image in css. If that's still true there should just be a classname like `&_image` that's used without nesting in the container.

KimberlyMunoz replied:
Does this also hold true for the 50-50 text and image organism?

jimmynotjim replied:
I don't believe so, I think those images are always going to be content and not a visual enhancement, but I could be wrong. @duelj @ajbush any input here?

anselmbradford replied:
Does this work? https://github.com/cfpb/cfgov-refresh/commit/8cebcf413e325c2c0c76a91dc1fc63558c7b35d7

jimmynotjim commented on cfgov/v1/preprocessed/css/molecules/image-and-text-25-75.less:
```diff
@@ -0,0 +1,93 @@
+                width: 150px;
+            } );
+        }
+        text-align: center;
```
Scope `text-align` and `margin-bottom` into a max-width mq to avoid resetting in the tablet min mq

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/dfaa799763fd6f6c99a87fb2749948eb6900f0f1

jimmynotjim commented on cfgov/v1/jinja2/v1/landing-page-2/index.html:
```diff
@@ -15,6 +16,12 @@
         {{ image_and_text_50_50.render() }}
+        {# TODO: Add stacked image and text organism with bottom padding. #}
+        <div style="padding-bottom: 60px;">
+            {{ image_and_text_25_75.render( 'landing-page-2/_template-prototype-image-and-text-25-75.html', '/static/img/educate.png' ) }}
```
This should use arguments instead of templates as well.

anselmbradford commented:
Squashed everything down. I can't test locally, but I relinquish this to @jimmynotjim @KimberlyMunoz @sebworks to merge.

jimmynotjim commented:
I had to rebase `flapjack` into this branch and then

- delete `cfgov/v1/static`
- run `pip uninstall django-sheerlike`
- run `pip install -r requirements/local.txt`

Everything is working, but I'm seeing `geee` in place of the images. Is that supposed to be there?

KimberlyMunoz commented on cfgov/jinja2/v1/_includes/molecules/image-and-text-25-75.html:
```diff
@@ -0,0 +1,54 @@
+       in which case the alt attribute is empty. #}
+    {% if settings.image_alt == '' %}
+        <div class="m-image-and-text-25-75_image"
+             >geee</div>
```
Here's the geee!

KimberlyMunoz replied:
Fixed in 75f5ea3

KimberlyMunoz commented:
Fixed the extra text and decorative images. Take another look @jimmynotjim?

jimmynotjim commented on cfgov/preprocessed/css/molecules/image-and-text-25-75.less:
```diff
@@ -0,0 +1,108 @@
+    }
+}
+
+h1.m-image-and-text-25-75_heading {
```
I just realized, there's two `_heading`s, anyone know what's up with that?

KimberlyMunoz replied:
I bet this is related to the fix with double-sized ems. I can look at if this can be cleaned up.

jimmynotjim replied:
I don't think so, he used the exposed `@font-size` in both instances and the spec actually says __not__ to use an `h1` because it'll compete with the page title.

KimberlyMunoz replied:
So is this just extra since there shouldn't ever be an h1 in this molecule.

jimmynotjim replied:
I think so. We can always check in with him on Mon about it.

KimberlyMunoz replied:
Sounds good! Although Monday is a CFPB holiday, so Tuesday it is.

anselmbradford replied:
Hm, yeah I don't think that's needed. It was originally to fix the heading bottom margin that appeared to not be necessary because of the defaults.

jimmynotjim replied:
Ah right, so this entire block can be removed and the margin bottom on the other `_heading` and `_body` can be removed, the `_link` is the only one that doesn't have a 30px bottom margin by default.

anselmbradford commented:
This is ready for review again.
- Normalized file naming between this and Image and Text 50/50.
- Added missing gray rule to both image and text molecules.
- Added missing changelog entry.
- Updated landing page to look more like mocks.

![screen shot 2015-10-13 at 6 25 53 pm](https://cloud.githubusercontent.com/assets/704760/10470119/a3b477ea-71d8-11e5-8120-c55a3de5a18f.png)

With gray rule turned on:
![screen shot 2015-10-13 at 6 24 28 pm](https://cloud.githubusercontent.com/assets/704760/10470122/a7a52fde-71d8-11e5-81e9-8ebe5165b855.png)

sebworks commented on cfgov/jinja2/v1/_includes/molecules/image-text-25-75.html:
```diff
@@ -0,0 +1,58 @@
+    </div>
+    <div class="m-image-text-25-75_content">
+        <h3 class="m-image-text-25-75_heading">{{ settings.heading }}</h3>
+
```
Unnecessary newline.

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/f390b9e69fff8c84414077be04934aacd8745803

sebworks commented on cfgov/jinja2/v1/_includes/molecules/image-text-50-50.html:
```diff
@@ -11,23 +11,27 @@
-   settings.title:        A string containing title of module
-   settings.image_path:   A string containing the url for the image
-   settings.image_alt:    A string containing alt text for screenreaders
+   settings:               An object with the following options for settings.
```
Should we make the settings documentation use the same format as the ` image_text_25_75` macro?

anselmbradford replied:
I'd be on board with that. For readability, I kind of prefer the indentation vs writing out `settings.` repeatedly, but looking across the codebase that's how it's done, so updated here to align with that  in https://github.com/cfpb/cfgov-refresh/commit/51c846244e362a15ee086f8bd11c6d3045cad1e5 (also fixes slight indentation issue in link blob).
Opened https://github.com/cfpb/cfgov-refresh/issues/1083 for it.

sebworks commented on cfgov/jinja2/v1/landing-page-1/index.html:
```diff
@@ -11,10 +12,42 @@
+            'image_path': 'http://placehold.it/300x300',
+            'image_alt': '',
+            'heading' : 'Placeholder',
+            'body': 'Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt.',
```
Can we just use the Jinja `lipsum` filter to create the text? Like 
```  {% set lipsumText = lipsum(1, false, 10, 25) | safe %} ```

http://jinja.pocoo.org/docs/dev/templates/#lipsum

anselmbradford replied:
Added in https://github.com/cfpb/cfgov-refresh/commit/8cd3c7c0377de184558b1543953032a2101bb97c

sebworks commented:
:+1:

jimmynotjim commented on cfgov/jinja2/v1/_includes/molecules/image-text-50-50.html:
```diff
@@ -11,23 +11,27 @@
-   settings.link_text:    A string for the text of the link/button call to action
+                           Will be 4:3 ratio by default.
+   settings.description:   A string for the description under the image
+   settings.is_button:     Boolean if the link should be a button. If false/blank,
```
Should this be `settings.link_is_button` to follow the rest of the link format?

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/f3595a300baf0a5429bfe71135bc596f0c866594

jimmynotjim commented on cfgov/jinja2/v1/_includes/molecules/half-width-link-blob.html:
```diff
@@ -8,18 +8,18 @@
 
-    settings.content: A string containing the content  to display below the
-                      heading (typically a paragraph).
+   settings.content: A string containing the content  to display below the
```
This should probably be updated to `body` to match the other macros. Can be done outside this PR.

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/71b9c454a263e013ed5733c8a402c8951e6f0c76

jimmynotjim commented on cfgov/jinja2/v1/landing-page-2/index.html:
```diff
@@ -34,6 +38,46 @@
+
+        <div class="content-l_col-1-2">
+            {{ image_text_50_50.render({
+                            'title' : 'Image and Text 50-50',
```
Un-indent here and check 80 char line lengths after.

anselmbradford replied:
Fixed in https://github.com/cfpb/cfgov-refresh/commit/434b591264048cb001bb2c998a6fbcd3f898f3d5

anselmbradford commented:
@jimmynotjim Ready to check again. Thanks!

jimmynotjim commented:
:+1:
